include/SampleLoader.h

```diff
@@ -22,27 +23,66 @@
  *
  */
 
-#ifndef LMMS_GUI_SAMPLE_LOADER_H
-#define LMMS_GUI_SAMPLE_LOADER_H
+#ifndef LMMS_SAMPLE_LOADER_H
+#define LMMS_SAMPLE_LOADER_H
 
+#include <QFileSystemWatcher>
+#include <QObject>
 #include <QString>
-#include <memory>
+#include <map>
 
 #include "SampleBuffer.h"
 #include "lmms_export.h"
 
-namespace lmms::gui {
-class LMMS_EXPORT SampleLoader
+namespace lmms {
+
+class LMMS_EXPORT SampleLoader : public QObject
 {
+	Q_OBJECT
 public:
-	static QString openAudioFile(const QString& previousFile = "");
-	static QString openWaveformFile(const QString& previousFile = "");
-	static std::shared_ptr<const SampleBuffer> createBufferFromFile(const QString& filePath);
-	static std::shared_ptr<const SampleBuffer> createBufferFromBase64(
-		const QString& base64, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	~SampleLoader() override = default;
+
+	using Source = SampleBuffer::Source;
+
+	enum class Cache
+	{
+		None,
+		Read,
+		ReadWrite
+	};
+
+	static auto inst() -> SampleLoader&;
+
+	static auto fromFile(const QString& filePath, Cache cache = Cache::ReadWrite)
+		-> std::shared_ptr<const SampleBuffer>;
+
+	static auto fromBase64(const QString& base64, int sampleRate, Cache cache = Cache::ReadWrite)
+		-> std::shared_ptr<const SampleBuffer>;
+
+	static auto fromBase64(const QString& base64, Cache cache = Cache::ReadWrite)
+		-> std::shared_ptr<const SampleBuffer>;
+
+private slots:
+	void removeFile(const QString& path);
+
 private:
-	static void displayError(const QString& message);
+	SampleLoader();
+
+	class AutoEvictor;
+
+	void add(const std::shared_ptr<const SampleBuffer>& buffer);
+	auto remove(const SampleBuffer& buffer) -> bool;
```

Comment:
The evictor only has a `SampleBuffer*` but has to call this method, so that's why it is the way it is. A const reference here is better than a pointer anyway, because passing a null pointer would be invalid and const references are like never-null pointers.

The only reason I didn't do the same thing for the `add()` method was because it needs to set the cache's weak pointer, and it's simple and efficient to just pass a shared pointer for that purpose.

Reply:
> The evictor only has a SampleBuffer* but has to call this method, so that's why it is the way it is.

Does the evictor for the shared pointers/weak pointers have to evict it from the cache? One no one is using a buffer, there would be no buffer in memory since we store `std::weak_ptr`'s, and its simpler to just leave it there in that case. When someone wants the data again, that entry would be filled in with an unexpried `std::weak_ptr`.

>  A const reference here is better than a pointer anyway, because passing a null pointer would be invalid and const references are like never-null pointers.

Huh? You're saying that I cant make a empty shared pointer (effectively `nullptr`) and pass that in to `add`? I thought you added the const reference because you did not want to increase the ref count when making the `weak_ptr` to be stored in the cache (however,  in that case, you could still take it by value and move it into the `weak_ptr`, which might be simpler to understand).